<a href="https://colab.research.google.com/github/mspendyala/data-engineering/blob/main/ExploratoryDataAnalysis/DetectAndRemoveOutliers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Detecting and Removing Outliers**

Detect and remove outliers using Z-score and IQR score

In [3]:
# Import needed libraries
import pandas as pd
from sklearn import datasets
from scipy import stats
import numpy as np
import sklearn

In [4]:
# Import california housing price dataset
# x is independent variables are many where as y is a dependent variable which is always one
california = sklearn.datasets.fetch_california_housing()
x = california.data
y = california.target
columns = california.feature_names

In [5]:
# Create the dataframe
california_df = pd.DataFrame(california.data)
california_df.columns = columns
california_df['prediction_target'] = y
california_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  prediction_target  
0    -122.23              4.526  
1    -122.22              3.585  
2    -122.24              3.521  
3    -122.25              3.413  
4    -122.25              3.422

**Using Z-Score**

What is Z-Score? 
How many standard deviations is the given value is this from the mean is called Z-Score. If Z-Score is zero then it is the same as the mean.

In [6]:
# Use Z-Score function defined in sciopy library to detect the outliers
# zscore could be above or below mean we are only taking absolute value

california_df_zscore = california_df
z_score = np.abs(stats.zscore(california_df))
print(z_score)

[[2.34476576 0.98214266 0.62855945 ... 1.05254828 1.32783522 2.12963148]
 [2.33223796 0.60701891 0.32704136 ... 1.04318455 1.32284391 1.31415614]
 [1.7826994  1.85618152 1.15562047 ... 1.03850269 1.33282653 1.25869341]
 ...
 [1.14259331 0.92485123 0.09031802 ... 1.77823747 0.8237132  0.99274649]
 [1.05458292 0.84539315 0.04021111 ... 1.77823747 0.87362627 1.05860847]
 [0.78012947 1.00430931 0.07044252 ... 1.75014627 0.83369581 1.01787803]]


8.3252	is 2.34476576 standard deviations away from mean of the MedINc column

In [7]:
#Define a threshold
threshold = 3
print(np.where(z_score>3))

(array([  131,   283,   409,   510,   511,   512,   514,   570,   576,
         710,   780,   799,   864,   865,   867,   869,   871,   922,
         923,   977,   985,   986,   995,  1010,  1021,  1024,  1024,
        1039,  1060,  1086,  1102,  1102,  1233,  1233,  1234,  1234,
        1235,  1235,  1238,  1238,  1239,  1239,  1240,  1240,  1541,
        1560,  1561,  1563,  1564,  1566,  1566,  1574,  1582,  1583,
        1586,  1591,  1593,  1602,  1617,  1621,  1636,  1637,  1642,
        1644,  1645,  1646,  1700,  1867,  1867,  1872,  1872,  1879,
        1889,  1889,  1910,  1910,  1911,  1911,  1912,  1912,  1913,
        1913,  1914,  1914,  1925,  1926,  1926,  1930,  1978,  1978,
        1979,  1979,  2025,  2119,  2213,  2294,  2311,  2392,  2392,
        2395,  2395,  2396,  2396,  2397,  2397,  2398,  2398,  2511,
        2511,  2776,  2826,  2963,  2969,  2975,  2978,  2999,  3004,
        3086,  3086,  3167,  3177,  3258,  3258,  3292,  3334,  3350,
        3350,  3364

In [8]:
# Print z-score of 17845[0]
print(z_score[131][0])

4.0694430501296885


In [13]:
#Remove the outliers
california_df_zscore_removed_outliers = california_df_zscore[((stats.zscore(california_df_zscore)) < 3).all(axis=1)]

print("The number of rows before outlier filtering was: ", california_df.shape)
print("The number of rows after outlier filtering was: ", california_df_zscore_removed_outliers.shape)


The number of rows before outlier filtering was:  (20640, 9)
The number of rows after outlier filtering was:  (19794, 9)


Hence, number of rows removed using outlier filter is 846

Using IQR Score - Interquartile range

In [15]:
# Calculate IQR
california_df_iqr = california_df
Q1 = california_df_iqr.quantile(0.25)
Q3 = california_df_iqr.quantile(0.75)
IQR = Q3 - Q1
display(IQR)

MedInc                 2.179850
HouseAge              19.000000
AveRooms               1.611665
AveBedrms              0.093447
Population           938.000000
AveOccup               0.852520
Latitude               3.780000
Longitude              3.790000
prediction_target      1.451250
dtype: float64

In [24]:
# Detect the outliers

display(california_df_iqr < (Q1 - 1.5 * IQR)) or (california_df_iqr > (Q3 + 1.5 * IQR))

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0       False     False     False      False       False     False     False   
1       False     False     False      False       False     False     False   
2       False     False     False      False       False     False     False   
3       False     False     False      False       False     False     False   
4       False     False     False      False       False     False     False   
...       ...       ...       ...        ...         ...       ...       ...   
20635   False     False     False      False       False     False     False   
20636   False     False     False      False       False     False     False   
20637   False     False     False      False       False     False     False   
20638   False     False     False      False       False     False     False   
20639   False     False     False      False       False     False     False   

       Longitude  prediction_target  
0          False              False  
1          False              False  
2          False              False  
3          False              False  
4          False              False  
...          ...                ...  
20635      False              False  
20636      False              False  
20637      False              False  
20638      False              False  
20639      False              False  

[20640 rows x 9 columns]

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0        True     False     False      False       False     False     False   
1        True     False     False      False       False     False     False   
2       False     False     False      False       False     False     False   
3       False     False     False      False       False     False     False   
4       False     False     False      False       False     False     False   
...       ...       ...       ...        ...         ...       ...       ...   
20635   False     False     False      False       False     False     False   
20636   False     False     False       True       False     False     False   
20637   False     False     False      False       False     False     False   
20638   False     False     False      False       False     False     False   
20639   False     False     False      False       False     False     False   

       Longitude  prediction_target  
0          False              False  
1          False              False  
2          False              False  
3          False              False  
4          False              False  
...          ...                ...  
20635      False              False  
20636      False              False  
20637      False              False  
20638      False              False  
20639      False              False  

[20640 rows x 9 columns]